esda/silhouettes.py

```diff
@@ -0,0 +1,406 @@
+    import sklearn.metrics.pairwise as skp
+    from sklearn.preprocessing import LabelEncoder
+    import pandas as pd
+    HAS_REQUIREMENTS = True
+except ImportError as e:
+    HAS_REQUIREMENTS = False
+
+def _raise_initial_error():
+    missing = []
+    try:
+        import sklearn
+    except ImportError:
+        missing.append('scikit-learn')
+    try:
+        import pandas
+    except ImportError:
+        missing.append('pandas')
+    raise ImportError("this function requires scikit-learn and "
+                      "pandas to be installed. Missing {}".format(','.join(missing)))
+
+def path_silhouette(data, labels, W, D=None, 
+                    metric=skp.euclidean_distances, 
+                    closest=False, return_nbfc=False, 
+                    return_nbfc_score=False,
+                    return_paths=False,
+                    directed=False):
+    """
+    Compute a path silhouette for all observations. 
+
+    Parameters
+    -----------
+    data    :   np.ndarray (N,P)
+                matrix of data with N observations and P covariates. 
+    labels  :   np.ndarray (N,)
+                flat vector of the L labels assigned over N observations. 
+    W       :   pysal.W object
+                spatial weights object reflecting the spatial connectivity 
+                in the problem under analysis
+    D       :   np.ndarray (N,N)
+                a precomputed distance matrix to apply over W. If passed, 
+                takes precedence over data, and data is ignored. 
+    metric  :   callable
+                function mapping the (N,P) data into an (N,N) dissimilarity matrix, 
+                like that found in scikit.metrics.pairwise or scipy.spatial.distance
+    closest :   bool
+                whether or not to consider the observation "connected" when it 
+                is first connected to the cluster, or considering the path cost
+                to transit through the cluster. If True, the path cost is assessed
+                between i and the path-closest j in each cluster. If False, the path
+                cost is assessed as the average of path costs between i and all j
+                in each cluster
+    return_nbfc     :   bool
+                        Whether or not to return the label of the next best fit 
+                        cluster
+    return_nbfc_score:  bool
+                        Whether or not to return the score of the next best fit 
+                        cluster. 
+    return_paths    :   bool
+                        Whether or not to return the matrix of shortest path 
+                        lengths after having computed them. 
+    directed    :   bool
+                    whether to consider the weights matrix as directed or undirected.
+                    If directed, asymmetry in the input W is heeded. If not,
+                    asymmetry is ignored. 
+
+    Returns
+    --------
+    An (N,) array of the clusters involved. 
```

Comment:
Maybe more details on what is returned could be helpful. Are these the labels for N observations, or N clusters?